src/training/graph_group.h

```diff
@@ -83,22 +85,20 @@ class AsyncGraphGroup : public GraphGroup {
         opt_->update(graphs_[0]);
       }
       else {
-        std::lock_guard<std::mutex> guard(sync_);
         // add instead of copy?
-        ThreadPool pool(devices_.size());
         std::vector< std::future<void> > results;
         int pos = 0;
         for (int idx = 0; idx < devices_.size(); idx++) {
-            results.push_back( std::async (std::launch::async, [=](int idx, int pos) {
+          std::thread([=](int idx, int pos) {
+            //individual mutex per-shard
+            std::lock_guard<std::mutex> guard( shardSync_[idx] );
             grads_[idx]->copyFrom( newGrads->subtensor(pos , grads_[idx]->size() ) );
             shardOpt_[idx]->update(params_[idx], grads_[idx]);
-          } , idx, pos));  
 
-          pos += shardSize_;
-        }
+            cudaStreamSynchronize(0);
+          } , idx, pos).detach();  
```

Comment:
how worse?

I tested with simple cout, there is no race condition between fetch and push. However, though, because each shard now has independent mutex, it is possible that gradientPush finish time between shards is different. i.e:

 -gpu K calls gradientPush
 -gradientPush runs 4 threads for each shard
 -shard 0 finished instantly, while other shards has to wait.
 -because independent mutex, other gpu can fetch from shard 0.

thus it is possible, that gradientFetch will fetch gradient form different timestamp. That sounds like messing up with convergence rate.